### PR TITLE
NO-ISSUE: Remove deleted hosts from metadata

### PR DESCRIPTION
**Pull request by javipolo**
We should clean the list of hosts while reading metadata by removing hosts that are deleted BEFORE installation starts

/cc @omertuc

### CODE REVIEW

**Comment by rccrdpccl** [comment]
This hopefully will be unnecessary when (if) we'll implement installation ID

/lgtm

File: tools/triage/add_triage_signature.py [hunk 1]
```diff
@@ -109,12 +109,26 @@ class FailedToGetInstallConfigException(Exception):
...
 
+def clean_metadata_json(md):
+    start_time = datetime.strptime(md["cluster"]["install_started_at"], "%Y-%m-%dT%H:%M:%S.%fZ")
+    hosts = []
+    for host in md["cluster"]["hosts"]:
+        if host["deleted_at"]:
+            delete_time = datetime.strptime(host["deleted_at"], "%Y-%m-%dT%H:%M:%S.%fZ")
+            if delete_time < start_time:
+                continue
+        hosts.append(host)
```
Comment (omertuc) [inline tools/triage/add_triage_signature.py:120]:
```suggestion
    md["cluster"]["hosts"] = [
        host
        for host in md["cluster"]["hosts"]
        if datetime.strptime(host["deleted_at"], "%Y-%m-%dT%H:%M:%S.%fZ") < start_time
    ]

```

File: tools/triage/add_triage_signature.py [hunk 2]
```diff
@@ -109,12 +109,26 @@ class FailedToGetInstallConfigException(Exception):
...
+    md["cluster"]["hosts"] = hosts
```
Comment (omertuc) [inline tools/triage/add_triage_signature.py:121]:
```suggestion
    md["cluster"]["hosts"] = [
        host
        for host in md["cluster"]["hosts"]
        if datetime.strptime(host["deleted_at"], "%Y-%m-%dT%H:%M:%S.%fZ") < start_time
    ]
```

**Comment by omertuc** [comment]
I like this idea, I like that it's the default behavior of the get metadata function, but I also want to make it optional, along with a new signature that lists all the deleted hosts

**Comment by omertuc** [comment]
Maybe just add a "virtual" field to the metadata called `deleted_hosts` that will contain the deleted hosts

File: tools/triage/add_triage_signature.py [hunk 3]
```diff
@@ -109,12 +109,29 @@ class FailedToGetInstallConfigException(Exception):
...
+        if host["deleted_at"] and datetime.strptime(host["deleted_at"], "%Y-%m-%dT%H:%M:%S.%fZ") < start_time
+    ]
+    md["cluster"]["hosts"] = [
+        host
+        for host in md["cluster"]["hosts"]
+        if not host["deleted_at"]
+        or host["deleted_at"]
+        and datetime.strptime(host["deleted_at"], "%Y-%m-%dT%H:%M:%S.%fZ") > start_time
+    ]
+    return md
```
Comment (omertuc) [inline tools/triage/add_triage_signature.py:126]:
```suggestion
from itertools import filterfalse

def clean_metadata_json(md):
    installation_start_time = dateutil.parser.isoparse(
        md["cluster"]["install_started_at"]
    )

    def host_deleted_before_installation_started(host):
        if (deleted_at := host.get("deleted_at")) is not None:
            return dateutil.parser.isoparse(deleted_at) < installation_start_time
        return False

    md["cluster"]["deleted_hosts"] = list(
        filter(host_deleted_before_installation_started, md["cluster"]["hosts"])
    )
    md["cluster"]["hosts"] = list(
        filterfalse(host_deleted_before_installation_started, md["cluster"]["hosts"])
    )

    return md
```

**Comment by javipolo** [comment]
I like the solution, pushed!

**Pull request merged**